src/main/kotlin/io/libp2p/core/protocol/ProtocolBinding.kt

```diff
@@ -0,0 +1,41 @@
+package io.libp2p.core.protocol
+
+import io.netty.channel.Channel
+import io.netty.channel.ChannelInitializer
+import java.util.concurrent.CompletableFuture
+
+/**
+ * A ProtocolBinding represents the entrypoint to a protocol.
+ *
+ * It provides metadata about the protocol (such as its protocol ID and matching heuristics), and activation logic.
+ *
+ * A protocol can act on a Connection (as is the case of StreamMuxers and SecureChannels), or it can be deployed on a
+ * Stream (e.g. user-land protocols such as pubsub, DHT, etc.)
+ */
+interface ProtocolBinding<TController> {
+    /**
+     * The protocol ID with which we'll announce this protocol to peers.
+     */
+    val announce: String
+
+    /**
+     * Matching heuristic to be used on inbound negotiations. If positive, it leads to activation of this protocol.
+     */
+    val matcher: ProtocolMatcher
+
+    /**
+     * Returns initializer for this protocol on the provided channel, together with an optional controller object.
+     */
+    fun initializer(): ProtocolBindingInitializer<TController>
```

Comment:
@raulk I've returned back to `ChannelInitializer` since with `activate(Channel)` I can do only `Pipeline.addLast()` while it's sometimes not desirable since another handlers could (and sometimes should) be installed at the end.

Reply:
That's weird. Do you get a runtime exception? Because the API should allow you to do `channel.pipeline().add*()`?

Reply:
This might not be exactly our use case, but how I see it: 
You may want in theory to initialize the channel with two handlers: `Secio` + `ApplicationHandler`. Secio handler itself adds and removes his own handlers. If it would do `addLast` then added handlers will be after the `ApplicationHandler` and this will break 'incapulation' of `Secio` handler. 
Thus instead of doing `addLast` is should either `replace` old handlers or call `addBefore/After` its own handlers. 
To achieve this the handler should know 'himself'. This is not possible with `activate(Channel)` approach.

Reply:
That would be ideally handled with kind of 'pipelines hierarchy' but this is not implemented in Netty